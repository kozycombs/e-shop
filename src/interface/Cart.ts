export interface CartProduct {
  productId: number;
  quantity: number;
}

export interface Cart {
  id: number;
  userId: number;
  date: string;
  products: CartProduct[];
}

export interface CartResponse {
  id: number;
}
