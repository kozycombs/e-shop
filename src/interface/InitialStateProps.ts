export interface InitialStateProps<T> {
  data: T;
  error: string;
}
