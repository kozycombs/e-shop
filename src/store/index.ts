import { configureStore } from "@reduxjs/toolkit";
import productReducer from "store/productSlice";
import cartReducer from "store/cartSlice";

export const store = configureStore({
  reducer: {
    products: productReducer,
    cart: cartReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;
