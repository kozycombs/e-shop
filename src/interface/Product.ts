import { Rating } from "interface/Rating";

export interface Product {
  id: number;
  title: string;
  price: number;
  description: string;
  category: string;
  image: string;
  rating: Rating;
}
