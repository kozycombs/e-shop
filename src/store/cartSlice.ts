import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { RootState } from ".";
import { API_BASE_URL } from "../constants";
import { InitialStateProps } from "../interface/InitialStateProps";
import { Cart, CartProduct } from "../interface/Cart";

export const initialState: InitialStateProps<Cart | null> = {
  loading: false,
  data: null,
  error: "",
};

export const cartSlice = createSlice({
  name: "cart",
  initialState,
  reducers: {
    updateCart: (state, action: { payload: Cart }) => {
      state.data = action.payload;
    },
    resetCart: () => {
      return { ...initialState };
    },
  },
  extraReducers(builder) {
    builder
      .addCase(addToCart.fulfilled, (state, action) => {
        state.data = action.payload;
      })
      .addCase(updateCartProduct.fulfilled, (state, action) => {
        state.data = action.payload;
      });
  },
});

export const { updateCart, resetCart } = cartSlice.actions;

export default cartSlice.reducer;

export const cartSelector = (state: RootState) => state.cart;

export const addToCart = createAsyncThunk(
  "cart/addToCart",
  async ({
    userId,
    date,
    products,
  }: {
    userId: number;
    date: string;
    products: CartProduct[];
  }): Promise<Cart> => {
    const response = await fetch(`${API_BASE_URL}/carts`, {
      method: "POST",
      body: JSON.stringify({
        userId,
        date,
        products,
      }),
    });
    const result = await response.json();
    return {
      id: Number(result.id),
      userId,
      date,
      products,
    };
  }
);

export const updateCartProduct = createAsyncThunk(
  "cart/updateCartProduct",
  async ({
    cartId,
    userId,
    date,
    products,
  }: {
    cartId: number;
    userId: number;
    date: string;
    products: CartProduct[];
  }): Promise<Cart> => {
    const response = await fetch(`${API_BASE_URL}/carts/${cartId}`, {
      method: "PUT",
      body: JSON.stringify({
        userId,
        date,
        products,
      }),
    });
    const result = await response.json();
    return {
      id: Number(result.id),
      userId,
      date,
      products,
    };
  }
);
