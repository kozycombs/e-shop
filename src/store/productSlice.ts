import { createSlice } from "@reduxjs/toolkit";
import { RootState } from ".";
import { Product } from "../interface/Product";

export const initialState: Product[] = [];

export const productSlice = createSlice({
  name: "products",
  initialState,
  reducers: {
    updateProducts: (state, action) => {
      state = action.payload;
    },
  },
});

export const { updateProducts } = productSlice.actions;

export default productSlice.reducer;

export const productsSelector = (state: RootState) => state.products;
