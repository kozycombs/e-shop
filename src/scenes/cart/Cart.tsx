import { FC, useState } from "react";
import {
  Alert,
  Box,
  Button,
  Card,
  CardContent,
  Divider,
  Grid,
  Snackbar,
  Typography,
} from "@mui/material";
import { useDispatch, useSelector } from "react-redux";
import { productsSelector } from "../../store/productSlice";
import { cartSelector, resetCart } from "../../store/cartSlice";
import CartItem from "../../components/cartItem/CartItem";
import { ToastMessage } from "../../interface/ToastMessage";
import { Link } from "react-router-dom";
import { APP_URL } from "../../constants";

const Cart: FC = () => {
  const dispatch = useDispatch();
  const { data: products } = useSelector(productsSelector);
  const { data: cart } = useSelector(cartSelector);
  const [toastMessage, setToastMessage] = useState<ToastMessage>({
    message: "",
    severity: "success",
  });
  const [orderConfirmed, setOrderConfirmed] = useState(false);
  const vertical = "top";
  const horizontal = "right";

  const resetToastMessage = () => {
    setToastMessage({ message: "", severity: "success" });
  };

  const getCartTotal = () => {
    return cart?.products.reduce((accumulator, cartItem) => {
      const cartProduct = products.find(
        (product) => product.id === cartItem.productId
      );
      if (!cartProduct) {
        return 0;
      }
      return (accumulator += cartProduct.price * cartItem.quantity);
    }, 0);
  };

  const handlePlaceOrder = () => {
    setOrderConfirmed((prev) => !prev);
    dispatch(resetCart());
  };

  if (orderConfirmed) {
    return (
      <Box sx={{ pt: 4 }}>
        <Typography variant="h1" sx={{ fontSize: 30, mb: 2 }}>
          Order Confirmation
        </Typography>
        <Divider sx={{ mb: 4 }} />
        <Typography>Your order was placed successfully.</Typography>
      </Box>
    );
  }

  return (
    <Box sx={{ pt: 4 }}>
      <Snackbar
        anchorOrigin={{ vertical, horizontal }}
        open={toastMessage?.message.length > 0}
        autoHideDuration={2000}
        onClose={resetToastMessage}
        key={vertical + horizontal}
      >
        <Alert
          onClose={resetToastMessage}
          severity={toastMessage?.severity}
          variant="filled"
          sx={{ width: "100%" }}
        >
          {toastMessage?.message}
        </Alert>
      </Snackbar>
      <Typography variant="h1" sx={{ fontSize: 30, mb: 2 }}>
        Your Basket
      </Typography>
      <Divider sx={{ mb: 4 }} />
      {!cart || cart?.products.length <= 0 ? (
        <Box sx={{ display: "flex", justifyContent: "space-around" }}>
          <Typography>
            Your basket is empty. View our{" "}
            <Link to={APP_URL.HOME} className="underline hover:no-underline">
              range of products
            </Link>
            .
          </Typography>
        </Box>
      ) : (
        <Grid container spacing={{ xs: 1, md: 4 }} columns={{ xs: 1, sm: 6 }}>
          <Grid item xs={1} sm={4}>
            {cart?.products.map((cartItem, index) => {
              const cartProduct = products.find(
                (product) => product.id === cartItem.productId
              );
              if (!cartProduct) {
                return "";
              }
              return (
                <CartItem
                  key={index}
                  cartItem={cartItem}
                  product={cartProduct}
                />
              );
            })}
          </Grid>
          <Grid item xs={1} sm={2}>
            <Card variant="outlined">
              <CardContent>
                <Typography variant="h2" sx={{ fontSize: 24, mb: 2 }}>
                  Summary
                </Typography>
                <Divider sx={{ mb: 4 }} />
                <Box
                  sx={{
                    display: "flex",
                    justifyContent: "space-between",
                    alignItems: "center",
                    mb: 1,
                  }}
                >
                  <Typography
                    component="p"
                    sx={{ fontSize: 18, fontWeight: 500, mb: 4 }}
                  >
                    Total
                  </Typography>
                  <Typography
                    component="p"
                    sx={{ fontSize: 18, fontWeight: 500, mb: 4 }}
                  >
                    &pound;{getCartTotal()}
                  </Typography>
                </Box>
                <Button
                  variant="contained"
                  size="medium"
                  color="success"
                  onClick={handlePlaceOrder}
                  sx={{ ml: "auto", display: "block" }}
                >
                  Place you order
                </Button>
              </CardContent>
            </Card>
          </Grid>
        </Grid>
      )}
    </Box>
  );
};

export default Cart;
