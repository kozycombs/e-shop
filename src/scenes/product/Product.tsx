import { FC, useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import {
  Alert,
  Box,
  Button,
  Divider,
  FormControl,
  Grid,
  MenuItem,
  Rating,
  Select,
  SelectChangeEvent,
  Snackbar,
  Typography,
} from "@mui/material";
import { fetchProduct, productsSelector } from "store/productSlice";
import { Product as ProductInterface } from "interface/Product";
import Spinner from "components/spinner/Spinner";
import { addToCart, cartSelector, updateCartProduct } from "store/cartSlice";
import { AppDispatch } from "store/index";
import { ToastMessage } from "interface/ToastMessage";
import { PRODUCT_QUATITY_COUNT } from "constants/index";
import ErrorBanner from "components/errorBanner/ErrorBanner";

const Product: FC = () => {
  const { id } = useParams();
  const dispatch = useDispatch<AppDispatch>();
  const { data: products } = useSelector(productsSelector);
  const { data: cart } = useSelector(cartSelector);
  const [loading, setLoading] = useState(false);
  const [product, setProduct] = useState<ProductInterface>();
  const [quantity, setQuantity] = useState(1);
  const [toastMessage, setToastMessage] = useState<ToastMessage>({
    message: "",
    severity: "success",
  });
  const [error, setError] = useState("");
  const userId = 1; // This is a dummy user id, since there is no logged in functionality.
  const vertical = "top";
  const horizontal = "right";

  useEffect(() => {
    const getProduct = async () => {
      try {
        const result = await dispatch(fetchProduct(Number(id))).unwrap();
        setProduct(result);
        setLoading(false);
      } catch (error) {
        setError("Failed to load our product. Please try again later");
        setLoading(false);
      }
    };

    // If product list is empty get product detail from API
    // otherwiser get product detail from cached product list.

    if (products.length <= 0) {
      setLoading(true);
      getProduct();
    } else {
      setLoading(true);
      const product =
        products.find((currentProduct) => currentProduct.id === Number(id)) ||
        null;
      if (product) {
        setProduct(product);
      }
      setLoading(false);
    }
  }, [id, products, dispatch]);

  const resetToastMessage = () => {
    setToastMessage({ message: "", severity: "success" });
  };

  const handleQuantityChange = (event: SelectChangeEvent) => {
    setQuantity(Number(event.target.value));
  };

  const handleAddToCart = async () => {
    if (!product) {
      return;
    }

    try {
      setLoading(true);
      const date = new Date().toISOString();

      // If cart is null add to cart else update with cart id.
      if (!cart) {
        const result = await dispatch(
          addToCart({
            userId,
            date,
            products: [{ productId: product?.id, quantity }],
          })
        ).unwrap();
        if (result.id) {
          setToastMessage({
            message: "Product added to cart successfully",
            severity: "success",
          });
        } else {
          setToastMessage({
            message: "Failed to add product to cart",
            severity: "error",
          });
        }
      } else {
        const productFoundIndex = cart.products.findIndex(
          (currentProduct) => currentProduct.productId === product.id
        );

        if (productFoundIndex > -1) {
          const updatedProduct = {
            ...cart.products[productFoundIndex],
            quantity: cart.products[productFoundIndex].quantity + quantity,
          };
          const otherProducts = cart.products.filter(
            (currentProduct) => currentProduct.productId !== product.id
          );
          const products = [...otherProducts, updatedProduct];
          const result = await dispatch(
            updateCartProduct({ cartId: cart.id, userId, date, products })
          ).unwrap();
          if (result.id) {
            setToastMessage({
              message: "Product added to cart successfully",
              severity: "success",
            });
            setQuantity(1);
          } else {
            setToastMessage({
              message: "Failed to add product to cart",
              severity: "error",
            });
          }
        } else {
          const products = [
            ...cart.products,
            { productId: product?.id, quantity },
          ];
          const result = await dispatch(
            updateCartProduct({ cartId: cart.id, userId, date, products })
          ).unwrap();
          if (result.id) {
            setToastMessage({
              message: "Product added to cart successfully",
              severity: "success",
            });
            setQuantity(1);
          } else {
            setToastMessage({
              message: "Failed to add product to cart",
              severity: "error",
            });
          }
        }
      }
      setLoading(false);
    } catch (error) {
      setError("Failed to add product to cart. Please try again later");
      setLoading(false);
    }
  };

  return (
    <Box sx={{ pt: 4 }}>
      {loading && (
        <Box className="flex justify-center">
          <Spinner />
        </Box>
      )}

      {error.length > 0 && <ErrorBanner message={error} />}

      <Snackbar
        anchorOrigin={{ vertical, horizontal }}
        open={toastMessage?.message.length > 0}
        autoHideDuration={2000}
        onClose={resetToastMessage}
        key={vertical + horizontal}
      >
        <Alert
          onClose={resetToastMessage}
          severity={toastMessage?.severity}
          variant="filled"
          sx={{ width: "100%" }}
        >
          {toastMessage?.message}
        </Alert>
      </Snackbar>

      <Grid container spacing={{ xs: 4, md: 3 }} columns={{ xs: 1, sm: 4 }}>
        <Grid item xs={1} sm={2} sx={{ marginBottom: 10 }}>
          <Box
            component="img"
            sx={{
              height: "auto",
              width: "100%",
              maxHeight: { xs: 400 },
              objectFit: "contain",
            }}
            alt={product?.title}
            src={product?.image}
          />
        </Grid>
        <Grid item xs={1} sm={2}>
          <Box></Box>
          <Typography
            variant="h1"
            sx={{ mb: 1, fontWeight: "400", fontSize: 24 }}
          >
            {product?.title}
          </Typography>
          <Box sx={{ display: "flex", alignItems: "end", mb: 2 }}>
            <Typography
              component="span"
              sx={{
                mr: 1,
                fontSize: 16,
              }}
            >
              {product?.rating.rate}
            </Typography>
            <Rating
              name="read-only"
              value={Number(product?.rating.rate)}
              readOnly
            />
            <Typography
              component="span"
              sx={{
                ml: 1,
                mt: 1,
                fontSize: 16,
                textDecoration: "underline",
              }}
            >
              {product?.rating.count} reviews
            </Typography>
          </Box>
          <Typography
            component="p"
            sx={{ fontSize: 24, fontWeight: 400, mb: 4 }}
          >
            &pound;{Number(product?.price).toFixed(2)}
          </Typography>
          <Box sx={{ display: "flex", alignItems: "center", mb: 4 }}>
            <FormControl sx={{ m: 1, minWidth: 120 }} size="small">
              <Select
                id="quantity"
                data-test="quantity"
                value={`${quantity}`}
                onChange={handleQuantityChange}
              >
                {Array(PRODUCT_QUATITY_COUNT)
                  .fill("_")
                  .map((_, index) => (
                    <MenuItem key={index} value={index + 1}>
                      {index + 1}
                    </MenuItem>
                  ))}
              </Select>
            </FormControl>
            <Button
              variant="contained"
              size="medium"
              disableElevation
              onClick={handleAddToCart}
              sx={{ textTransform: "none" }}
            >
              Add to Cart
            </Button>
          </Box>
          <Divider />
          <Typography
            component="p"
            sx={{
              mb: 2,
              mt: 4,
              fontSize: 16,
            }}
          >
            {product?.description}
          </Typography>
        </Grid>
      </Grid>
    </Box>
  );
};

export default Product;
