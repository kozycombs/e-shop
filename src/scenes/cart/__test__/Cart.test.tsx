import { Provider } from "react-redux";
import { render, screen, configure } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { updateProducts } from "../../../store/productSlice";
import { store } from "../../../store";
import { Product } from "../../../interface/Product";
import CartPage from "../Cart";
import { resetCart, updateCart } from "../../../store/cartSlice";
import { Cart } from "../../../interface/Cart";

describe("<CartPage />", () => {
  configure({
    testIdAttribute: "data-test",
  });

  const product: Product = {
    id: 4,
    title:
      "Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket",
    price: 29.95,
    description:
      "100% POLYURETHANE(shell) 100% POLYESTER(lining) 75% POLYESTER 25% COTTON (SWEATER), Faux leather material for style and comfort / 2 pockets of front, 2-For-One Hooded denim style faux leather jacket, Button detail on waist / Detail stitching at sides, HAND WASH ONLY / DO NOT BLEACH / LINE DRY / DO NOT IRON",
    category: "women's clothing",
    image: "https://fakestoreapi.com/img/81XH0e8fefL._AC_UY879_.jpg",
    rating: {
      rate: 2.9,
      count: 340,
    },
  };

  const cartData: Cart = {
    id: 11,
    userId: 1,
    date: "2024-06-03T00:27:58.741Z",
    products: [
      {
        productId: 4,
        quantity: 2,
      },
    ],
  };

  beforeAll(() => {
    store.dispatch(updateProducts([product]));
    store.dispatch(updateCart(cartData));
  });

  it("should render the component", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <CartPage />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the correct cart total price", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <CartPage />
        </BrowserRouter>
      </Provider>
    );

    const expectedCartTotal = cartData.products.reduce((acc, cartProduct) => {
      if (cartProduct.productId !== product.id) {
        return 0;
      }
      return (acc += product.price * cartProduct.quantity);
    }, 0);
    const cartTotal = screen.getByTestId("cartTotal");
    expect(cartTotal.textContent).toBe(
      `Â£${Number(expectedCartTotal).toFixed(2)}`
    );
  });

  it("should have a Place your order", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <CartPage />
        </BrowserRouter>
      </Provider>
    );

    const placeOrderButton = screen.getByRole("button");
    expect(placeOrderButton.textContent).toBe("Place your Order");
  });

  it("should render Your basket is empty message when no product item is in the cart", () => {
    store.dispatch(resetCart()); // Reset products list in redux store
    render(
      <Provider store={store}>
        <BrowserRouter>
          <CartPage />
        </BrowserRouter>
      </Provider>
    );

    const productTitle = screen.getByText(/Your basket is empty./i);
    expect(productTitle).toBeInTheDocument();
  });
});
