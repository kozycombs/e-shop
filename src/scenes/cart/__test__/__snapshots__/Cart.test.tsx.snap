// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartPage /> should render the component 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-xuhmcu"
  >
    <h1
      class="MuiTypography-root MuiTypography-h1 css-1vcc1nn-MuiTypography-root"
    >
      Your Basket
    </h1>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-1t2s6c-MuiDivider-root"
    />
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-spacing-md-4 css-pt1vvu-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-sm-4 css-fqifuh-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-spacing-md-3 css-1njvgv6-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-sm-1 css-w9nz8z-MuiGrid-root"
          >
            <a
              href="/product/4"
            >
              <img
                alt="Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket"
                class="MuiBox-root css-10514to"
                src="https://fakestoreapi.com/img/81XH0e8fefL._AC_UY879_.jpg"
              />
            </a>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-sm-3 css-1rn35ll-MuiGrid-root"
          >
            <a
              href="/product/4"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1xh9tsm-MuiTypography-root"
              >
                Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-15okrha-MuiTypography-root"
              >
                quantity: 2
              </p>
            </a>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-sm-1 css-w9nz8z-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-0"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-jcjl1m-MuiTypography-root"
              >
                £59.9
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-sm-2 css-oe53i6-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root MuiTypography-h2 css-l7ef1q-MuiTypography-root"
            >
              Summary
            </h2>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth css-1t2s6c-MuiDivider-root"
            />
            <div
              class="MuiBox-root css-pcqfzb"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1divac3-MuiTypography-root"
              >
                Total
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1divac3-MuiTypography-root"
                data-test="cartTotal"
              >
                £59.90
              </p>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess css-1n7bogu-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Place your Order
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
